File: ResourceManager/ViewModels/Activities/ActivityPopupViewModel.cs
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using NetCore.Mvvm.ViewModels;
using NetCore.Security;
using NetCore.Validators;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class ActivityPopupViewModel : PopupViewModelBase
    {
        private readonly IDataServiceManager _dataServiceManager;
        private int ID;
        private bool isEdit;

        public ActivityPopupViewModel(IDataServiceManager dataServiceManager, IWindowManager windowManager)
        {
            _dataServiceManager = dataServiceManager;
            isEdit = false;

            AddValidationRule(() => Title, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Title)));
            AddValidationRule(() => Metric, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Metric)));
            AddValidationRule(() => EndTime, x => Duration > 0, nameof(EndTime) + " must greater than start time.");
            AddValidationRule(() => Quantity, x => Quantity > 0, nameof(Quantity) + " must greater than 0.");

            DateTime now = DateTime.Now;
            Date = now.Date;
            StartTime = new DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, 0);
            EndTime = new DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, 0);

            SaveCommand = new AsyncRelayCommand(OnSaveAsync, CanSave);
        }

        private bool CanSave()
        {
            return !HasErrors;
        }
        public void SetActivity(Activity activity)
        {
            isEdit = true;
            ID = activity.ID;
            Title = activity.Title;
            Duration = activity.Duration;
            Quantity = activity.Quantity;
            Metric = activity.Metric;
            Date = activity.Date;
            StartTime = activity.StartTime;
            EndTime = activity.EndTime;
            Output = activity.Output;
            Note = activity.Note;
        }
        private async Task OnSaveAsync()
        {
            ValidateAllRules();

            if (!HasErrors)
            {
                try
                {
                    var act = new Activity()
                    {
                        Title = Title,
                        Duration = Duration,
                        Quantity = Quantity,
                        Metric = Metric,
                        Date = Date.Date,
                        StartTime = StartTime,
                        EndTime = EndTime,
                        Output = Output,
                        Note = Note,
                        UserID = AuthManager.User.Identity.ID
                    };

                    if (isEdit)
                    {
                        await _dataServiceManager.ActivityDataService.Update(ID, act);

                        Result = PopupResult.OK;
                    }
                    else
                    {
                        await _dataServiceManager.ActivityDataService.Create(act);

                        Result = PopupResult.OK;
                    }
                }
                catch (Exception ex)
                {
                    windowManager.ShowMessageBox(ex.Message,System.Windows.MessageBoxButton.OK,System.Windows.MessageBoxImage.Error);
                }
                OnClose();
            }
        }

        public string Title
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public int Quantity
        {
            get => GetProperty<int>();
            set => SetProperty(value);
        }
        public string Metric
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public DateTime Date
        {
            get => GetProperty<DateTime>();
            set
            {
                if (SetProperty(value))
                {
                    StartTime = new DateTime(value.Year, value.Month, value.Day, StartTime.Hour, StartTime.Minute, 0);
                    EndTime = new DateTime(value.Year, value.Month, value.Day, EndTime.Hour, EndTime.Minute, 0);
                }
            }
        }
        public int Duration
        {
            get => GetProperty<int>();
            set => SetProperty(value);
        }
        public DateTime StartTime
        {
            get => GetProperty<DateTime>();
            set
            {
                if(SetProperty(value))
                {
                    Duration = (int)(EndTime - StartTime).TotalMinutes;
                }
            }
        }
        public DateTime EndTime
        {
            get => GetProperty<DateTime>();
            set
            {
                if (SetProperty(value))
                {
                    Duration = (int)(EndTime - StartTime).TotalMinutes;
                }
            }
        }
        public string Output
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string Note
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public ICommand SaveCommand { get; }
    }
}

File: ResourceManager/ViewModels/Activities/ActivityPopupViewModel.vb
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Mvvm.ViewModels
Imports VBNetCore.Security
Imports VBNetCore.Validators
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions
Imports System
Imports System.Threading.Tasks
Imports System.Windows.Input

Namespace ResourceManager.ViewModels
    Public Class ActivityPopupViewModel
        Inherits PopupViewModelBase
        Private ReadOnly _dataServiceManager As IDataServiceManager
        Private ID As Integer
        Private isEdit As Boolean

        Public Sub New(ByVal dataServiceManager As IDataServiceManager, ByVal windowManager As IWindowManager)
            _dataServiceManager = dataServiceManager
            isEdit = False

            AddValidationRule(Function() Title, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(ActivityPopupViewModel.Title)))
            AddValidationRule(Function() Metric, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(ActivityPopupViewModel.Metric)))
            AddValidationRule(Function() EndTime, Function() Duration > 0, NameOf(ActivityPopupViewModel.EndTime) & " must greater than start time.")
            AddValidationRule(Function() Quantity, Function() Quantity > 0, NameOf(ActivityPopupViewModel.Quantity) & " must greater than 0.")

            Dim now = Date.Now
            [Date] = now.Date
            StartTime = New DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, 0)
            EndTime = New DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, 0)

            SaveCommand = New AsyncRelayCommand(AddressOf OnSaveAsync, AddressOf CanSave)
        End Sub

        Private Function CanSave() As Boolean
            Return Not HasErrors
        End Function
        Public Sub SetActivity(ByVal activity As Activity)
            isEdit = True
            ID = activity.ID
            Title = activity.Title
            Duration = activity.Duration
            Quantity = activity.Quantity
            Metric = activity.Metric
            [Date] = activity.Date
            StartTime = activity.StartTime
            EndTime = activity.EndTime
            Output = activity.Output
            Note = activity.Note
        End Sub
        Private Async Function OnSaveAsync() As Task
            ValidateAllRules()

            If Not HasErrors Then
                Try
                    Dim act = New Activity() With {
    .Title = Title,
    .Duration = Duration,
    .Quantity = Quantity,
    .Metric = Metric,
    .[Date] = [Date].Date,
    .StartTime = StartTime,
    .EndTime = EndTime,
    .Output = Output,
    .Note = Note,
    .UserID = AuthManager.User.Identity.ID
}

                    If isEdit Then
                        Await _dataServiceManager.ActivityDataService.Update(ID, act)

                        Result = PopupResult.OK
                    Else
                        Await _dataServiceManager.ActivityDataService.Create(act)

                        Result = PopupResult.OK
                    End If
                Catch ex As Exception
                    windowManager.ShowMessageBox(ex.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error)
                End Try
                OnClose()
            End If
        End Function

        Public Property Title As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property Quantity As Integer
            Get
                Return GetProperty(Of Integer)()
            End Get
            Set(ByVal value As Integer)
                SetProperty(value)
            End Set
        End Property
        Public Property Metric As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property [Date] As Date
            Get
                Return GetProperty(Of Date)()
            End Get
            Set(ByVal value As Date)
                If SetProperty(value) Then
                    StartTime = New DateTime(value.Year, value.Month, value.Day, StartTime.Hour, StartTime.Minute, 0)
                    EndTime = New DateTime(value.Year, value.Month, value.Day, EndTime.Hour, EndTime.Minute, 0)
                End If
            End Set
        End Property
        Public Property Duration As Integer
            Get
                Return GetProperty(Of Integer)()
            End Get
            Set(ByVal value As Integer)
                SetProperty(value)
            End Set
        End Property
        Public Property StartTime As Date
            Get
                Return GetProperty(Of Date)()
            End Get
            Set(ByVal value As Date)
                If SetProperty(value) Then
                    Duration = CInt((EndTime - StartTime).TotalMinutes)
                End If
            End Set
        End Property
        Public Property EndTime As Date
            Get
                Return GetProperty(Of Date)()
            End Get
            Set(ByVal value As Date)
                If SetProperty(value) Then
                    Duration = CInt((EndTime - StartTime).TotalMinutes)
                End If
            End Set
        End Property
        Public Property Output As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property Note As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public ReadOnly Property SaveCommand As ICommand
    End Class
End Namespace
